core/l1_fee_context.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"bytes"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/predeploys"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var big10 = big.NewInt(10)
+
+var (
+	L1BaseFeeSlot = common.BigToHash(big.NewInt(1))
+	OverheadSlot  = common.BigToHash(big.NewInt(3))
+	ScalarSlot    = common.BigToHash(big.NewInt(4))
+	DecimalsSlot  = common.BigToHash(big.NewInt(5))
+)
+
+// calculateL1GasUsed returns the gas used to include the transaction data in
+// the calldata on L1
+func calculateL1GasUsed(data []byte, overhead *big.Int) *big.Int {
+	var zeroes uint64
+	var ones uint64
+	for _, byt := range data {
+		if byt == 0 {
+			zeroes++
+		} else {
+			ones++
+		}
+	}
+
+	zeroesGas := zeroes * params.TxDataZeroGas
```

Comment:
These params are likely going to be updated in a future hardfork - thinking of a way to make it so that the values can be modularized better such that chainconfig can determine what the values are

Reply:
Is this a blocker? cc @trianglesphere

Reply:
hmm, not fully b/c we control the upgrade cycle, but the `params.TxDataZeroGas` can change based on the chain config + HF schedule. We'd have to make sure that it's pulled out to a separate place if the parameter is every modified again

Reply:
Would it help to copy this param to Optimism config?

Reply:
Noted. Can think about this after this PR.